js/subprocess.ts

```diff
@@ -0,0 +1,136 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+import * as dispatch from "./dispatch";
+import * as flatbuffers from "./flatbuffers";
+import * as msg from "gen/msg_generated";
+import { assert, unreachable } from "./util";
+import { close, File } from "./files";
+import { ReadCloser, WriteCloser } from "./io";
+
+/** How to handle subsubprocess stdio.
+ *
+ * "inherit" The default if unspecified. The child inherits from the
+ * corresponding parent descriptor.
+ *
+ * "piped"  A new pipe should be arranged to connect the parent and child
+ * subprocesses.
+ *
+ * "null" This stream will be ignored. This is the equivalent of attaching the
+ * stream to /dev/null.
+ */
+export type SubprocessStdio = "inherit" | "piped" | "null";
+
+// TODO Maybe extend VSCode's 'CommandOptions'?
+// tslint:disable-next-line:max-line-length
+// See https://code.visualstudio.com/docs/editor/tasks-appendix#_schema-for-tasksjson
+export interface RunOptions {
+  args: string[];
+  cwd?: string;
+  stdout?: SubprocessStdio;
+  stderr?: SubprocessStdio;
+  stdin?: SubprocessStdio;
+}
+
+export class Subprocess {
+  readonly rid: number;
+  readonly pid: number;
+  readonly stdin?: WriteCloser;
+  readonly stdout?: ReadCloser;
+  readonly stderr?: ReadCloser;
+
+  // @internal
+  constructor(res: msg.RunRes) {
+    this.rid = res.rid();
+    this.pid = res.pid();
+
+    if (res.stdinRid() > 0) {
+      this.stdin = new File(res.stdinRid());
+    }
+
+    if (res.stdoutRid() > 0) {
+      this.stdout = new File(res.stdoutRid());
+    }
+
+    if (res.stderrRid() > 0) {
+      this.stderr = new File(res.stderrRid());
+    }
+  }
+
+  async status(): Promise<SubprocessStatus> {
+    return await runStatus(this.rid);
+  }
+
+  close(): void {
+    close(this.rid);
+  }
+}
+
+export interface SubprocessStatus {
```

Comment:
This type doesn't properly capture that either `code` or `signal` will have a value.
Thus the user won't be able to do:

```
declare function handleExitCode(code: number);
declare function handleExitSignal(sig: number);

if (status.code != null) {
  handleExitCode(status.code);
} else {
  handleExitSignal(status.signal);
}
```

Reply:
That code works currently.

Reply:
I concede -- It works because we using typescript in sloppy mode.
OK for now then.